proc/process.c

```diff
@@ -877,93 +875,76 @@ int process_load(process_t *process, vm_object_t *o, off_t base, size_t size, vo
-				}
-
-				/* Don't modify ELF file! */
-				if (((ptr_t)relptr >= (ptr_t)base) && ((ptr_t)relptr < ((ptr_t)base + size))) {
-					++badreloc;
-					continue;
-				}
-
-				sym = (Elf32_Sym *)(symTab + (ELF32_R_SYM(rela->r_info) * sizeof(Elf32_Sym)));
-
-				/* Write addend to the address */
-				*(char **)relptr = (char *)(sym->st_value + rela->r_addend);
-
-				if (process_relocate(reloc, relocsz, relptr) < 0) {
-					return -ENOEXEC;
-				}
-			}
+		/* strncmp as there may be multiple .rela.* or .rel.* sections for different sections.  */
+		if (hal_strncmp(&snameTab[shdr->sh_name], ".rela.", 6) == 0) {
+			isrela = 1;
 		}
-	}
-#else
-	/* Perform data, init_array and fini_array relocation */
-	for (i = 0, shdr = (void *)((char *)ehdr + ehdr->e_shoff); i < ehdr->e_shnum; i++, shdr++) {
-		/* strncmp as there may be multiple .rel.* sections for different sections. */
-		if (hal_strncmp(&snameTab[shdr->sh_name], ".rel", 4) != 0) {
-			continue;
+		else if (hal_strncmp(&snameTab[shdr->sh_name], ".rel.", 5) == 0) {
+			isrela = 0;
 		}
-
-		if ((shdr->sh_size == 0) || (shdr->sh_entsize == 0)) {
+		else {
 			continue;
 		}
 
-		for (j = 0; j < shdr->sh_size / shdr->sh_entsize; ++j) {
-			rel = (void *)((ptr_t)shdr->sh_offset + (ptr_t)ehdr + (j * shdr->sh_entsize));
-			reltype = ELF32_R_TYPE(rel->r_info);
+		for (j = 0; j < (shdr->sh_size / shdr->sh_entsize); ++j) {
+			/* Valid for both Elf32_Rela and Elf32_Rel, due to correct size being stored in shdr->sh_entsize. */
+			/* For .rel. section make sure not to access addend field! */
+			hal_memcpy(&rela, (Elf32_Rela *)((ptr_t)shdr->sh_offset + (ptr_t)ehdr + (j * shdr->sh_entsize)), shdr->sh_entsize);
+			reltype = ELF32_R_TYPE(rela.r_info);
 
-			if (reltype == R_ARM_ABS32 || reltype == R_ARM_TARGET1) {
-				relptr = (void *)rel->r_offset;
+			if (hal_isRelReloc(reltype) == 0) {
+				continue;
+			}
 
-				if (process_relocate(reloc, relocsz, (char **)&relptr) < 0) {
-					return -ENOEXEC;
-				}
+			relptr = (void *)rela.r_offset;
+			if (process_relocate(reloc, relocsz, (char **)&relptr) < 0) {
+				return -ENOEXEC;
+			}
 
-				/* Don't modify ELF file! */
-				if (((ptr_t)relptr >= (ptr_t)base) && ((ptr_t)relptr < ((ptr_t)base + size))) {
-					++badreloc;
-					continue;
-				}
+			/* Don't modify ELF file! */
+			if (((ptr_t)relptr >= (ptr_t)base) && ((ptr_t)relptr < ((ptr_t)base + size))) {
+				++badreloc;
+				continue;
+			}
 
-				if (process_relocate(reloc, relocsz, relptr) < 0) {
+			if (symTab == NULL) {
+				/* Accept files stripped from symtab if only SYM_UND is present, eg. files stripped with strip.py */
+				if (ELF32_R_SYM(rela.r_info) != 0) {
 					return -ENOEXEC;
 				}
+				symval = 0;
+			}
+			else {
+				symval = symTab[ELF32_R_SYM(rela.r_info)].st_value;
+			}
+
+			if (isrela == 0) {
+				/* If Rel is used addend stored is in the location. */
+				*(char **)relptr += symval;
+			}
+			else {
+				*(char **)relptr = (char *)(symval + rela.r_addend);
```

Comment:
I don't know if we expect any more relocations in kernel, but it's worth noting that different relocation types might need different calculations, so maybe it would be better to create some per-arch function like `hal_doReloc` which would do appropriate calculation for each relocation type (and then [this comment](https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/611/files#r1816169695) would be no issue)

Reply:
I prefer the approach in which we group relocations, independent of the architecture, based on how they are calculated.
Looking at the list of relocations on both architecture this should be doable.